### PR TITLE
boost serialization replaced by cereal

### PR DESCRIPTION
**Pull request by kenshi84**
Issue #1430 

It was quite confusing to understand how I should go about this, because there are *three* different kinds of serialization schemes co-existing in the codebase:

- `epee::serialization::kv_serialize(...)`
- `::serializaization::serialize(...)` which is called from `parse_binary` and `dump_binary` in serialization/binary_utils.h
- `boost::serialization::serialize(...)` which is called via `boost::archive::binary_{i|o}archive`

I replaced boost serializer by cereal except for p2p/net_peerlist.h which uses `boost::multi_index` which is not supported by cereal.

I confirmed `cereal::PortableBinary{Input|Output}Archive` to work compatibly across Windows and Mac.

Hope I didn't bring a mess to the codebase :D

### CODE REVIEW

**Comment by NanoAkron** [comment]
Totally naive question, in no way a criticism, what do you think using cereal will bring to the table that boost doesn't?

**Comment by kenshi84** [comment]
@NanoAkron It brings binary compatibility. So you can simply move around wallet cache files to other computers and open from there without worrying about the specific boost version used for build. It also seems to make the file size a bit smaller (48MB -> 38MB).

**Comment by moneromooo-monero** [comment]
This is interesting. I assume it can't read current wallets ?

**Comment by NanoAkron** [comment]
Portability is indeed very interesting, as is reducing the 3 separate serialisation mechanisms in the codebase into just one.

**Comment by anonimal** [comment]
>https://github.com/monero-project/monero/pull/1435/commits/697958279e72b86cd874d82b40faeff62a4bed5e

For external dependencies that clearly have a git source, why not use a git submodule instead of adding all these files? (aside from repo clutter, a submodule would ease upgrading).

>worrying about the specific boost version

I've read #1430, but is this still the strongest use-case for a relatively unpopular (compared to boost) library?

> it does not perform the same level of object tracking as other serialization libraries such as Boost. As a consequence of this raw pointers and references are not supported

Has anyone reviewed this and [other caveats](https://uscilab.github.io/cereal/)?

**Comment by fluffypony** [comment]
Please rebase

**Comment by kenshi84** [comment]
@moneromooo-monero: Of course not, I suppose, since the whole serialization code changed.

@NanoAkron: 
> Slight conspiracy theorist in me does wonder whether a code library released by a massive Chinese corporation might have subtle backdoors for their government agents...

It's released from University of Southern California:
https://github.com/USCiLab
http://ilab.usc.edu/
Where did you get the idea of the library being released by a massive Chinese corporation?

@fluffypony: I'll do so shortly.

@anonimal: 
> For external dependencies that clearly have a git source, why not use a git submodule instead of adding all these files? (aside from removing repo clutter, a submodule would ease upgrading).

I chose not to use git submodule since all third-party libraries in the monero repo (e.g. [RapidJSON](https://github.com/miloyip/rapidjson/)) were not using submodules, and starting to use it would force the users to add `--recursive` to the git command. Do we agree that submodules are better?

> I've read #1430, but is this still the strongest use-case for a relatively unpopular (compared to boost) library?

Yes, and in my opinion it's an important one; imagine a user losing access to wallet seed and only having the wallet cache. Being unable to open it due to wrong boost version and wrong platform would be a nightmare. Although cereal is less popular than boost, it's been popular for a couple of years now (and I've been using it) with quite many websites mentioning it including [51 questions on StackExchange](https://stackoverflow.com/questions/tagged/cereal).

> Has anyone reviewed this and other caveats?

Being unable to handle raw pointers (but able to handle `std::shared_ptr` etc) doesn't seem like a caveat to me. The library is fairly small and self-contained, so it'd be not much of a big deal to audit the code, I suppose.

**Comment by NanoAkron** [comment]
// Tencent is pleased to support the open source community by making RapidJSON available.
//
// Copyright (C) 2015 THL A29 Limited, a Tencent company, and Milo Yip. All rights reserved.

**Comment by anonimal** [comment]
@kenshi84 

>Do we agree that submodules are better?

Yes, and when I or someone gets to resolving https://github.com/monero-project/monero-core/issues/62; everyone will be pulling recursively or initializing separately *but*, for this PR, I just wanted to get the question out there.

>doesn't seem like a caveat to me.

It's a *huge* caveat (as are the others) depending on monero's (or anyone's) implementation! I haven't looked at the impl though (but I know you all have) so I just wanted to be sure that potential disasters are avoided.

**Comment by kenshi84** [comment]
@NanoAkron 
thanks, I didn't know about this Tencent company and overlooked that RapidJSON was from them. It sounds scary, but the monero repo seems to be already using it, so if this library is doomed bad, we should replace some parts of monero using it. The library seems fairly legit and easy to audit, though.

@anonimal 
> everyone will be pulling recursively or initializing separately but, for this PR, I just wanted to get the question out there.

I see. So for this PR is it OK to not have it as a submodule?

> It's a huge caveat (as are the others) depending on monero's (or anyone's) implementation! I haven't looked at the impl though (but I know you all have) so I just wanted to be sure that potential disasters are avoided.

From my understanding, all the relevant boost serialization codes are in:
- cryptonote_boost_serialization.h
- account_boost_serialization.h
(while some other files are just unnecessary; e.g. blockchain_storage_boost_serialization.h was never included anywhere)

I never encountered any instances of serializing raw pointers. Raw pointers seem to be seldom (or, never?) used as class members, as far as I saw.
What do you mean by "as are the others"?

BTW, I realized that the `parse_binary` and `dump_binary` actually offer binary compatibility; perhaps I should simply replace boost serialization with them.

**Comment by kenshi84** [comment]
@fluffypony rebased

**Comment by anonimal** [comment]
@kenshi84 

>So for this PR is it OK to not have it as a submodule?

Fine by me until submodule additions are done elsewhere, but that's just me.

>What do you mean by "as are the others"?

Everything in ["Advanced Topics"](https://uscilab.github.io/cereal/index.html). They may be completely irrelevant, I don't know since I'm not working on this code (that's why I asked if [anyone had reviewed](https://github.com/monero-project/monero/pull/1435#issuecomment-266492282)).

File: src/wallet/wallet2.cpp [hunk 1]
```diff
@@ -2289,15 +2287,15 @@ void wallet2::load(const std::string& wallet_, const std::string& password)
 
       std::stringstream iss;
       iss << cache_data;
-      boost::archive::binary_iarchive ar(iss);
+      cereal::PortableBinaryInputArchive ar(iss);
```
Comment (vtnerd) [inline src/wallet/wallet2.cpp]: The binary formats are not compatible, so what happens when a user with an existing wallet opens the file? A conversion tool or supporting multiple input formats may be needed. Output should be fine (transparent upgrade).

File: src/cryptonote_core/tx_pool.h [hunk 2]
```diff
@@ -515,10 +514,9 @@ namespace boost
         return;
       ar & td.kept_by_block;
     }
-  }
 }
-BOOST_CLASS_VERSION(cryptonote::tx_memory_pool, CURRENT_MEMPOOL_ARCHIVE_VER)
-BOOST_CLASS_VERSION(cryptonote::tx_memory_pool::tx_details, CURRENT_MEMPOOL_TX_DETAILS_ARCHIVE_VER)
+CEREAL_CLASS_VERSION(cryptonote::tx_memory_pool, CURRENT_MEMPOOL_ARCHIVE_VER)
```
Comment (vtnerd) [inline src/cryptonote_core/tx_pool.h]: These formats are unlikely to be compatible.

File: src/wallet/wallet2.cpp (hunk 1 above)
Comment (kenshi84) [inline src/wallet/wallet2.cpp]: Thank you so much for reviewing, that was helpful!
The revised version handles the old boost format as well.

**Comment by kenshi84** [comment]
@vtnerd @moneromooo-monero 
I fixed the code so that it can fall back to the old boost serialization when the de-serialization with cereal fails.

I tried to rebase the commits but couldn't figure out how; I made a merge commit in-between with manual conflict resolution for "GUI address book (5df9287)" which seems to prevent the rebasing.

**Comment by fluffypony** [comment]
@kenshi84 if it's an older format does it just handle it as-is with the fallback, or does it overwrite it with a cereal-serialised version on save?

**Comment by kenshi84** [comment]
@fluffypony currently, it overwrites it with the cereal code. I thought this behavior makes sense. Perhaps should it create a copy of the old cache file just in case?

**Comment by kenshi84** [comment]
@fluffypony Not it copies the old file as mywallet_old.

**Comment by NanoAkron** [comment]
As an aside, do you know whether the cereal team are working to provide functionality equivalent to boost::multi_index so that we only need a single serialisation solution?

File: src/wallet/wallet2.cpp [hunk 3]
```diff
@@ -2313,16 +2311,36 @@ void wallet2::load(const std::string& wallet_, const std::string& password)
...
       ar >> *this;
+      if (m_account_public_address.m_spend_public_key != m_account.get_keys().m_account_address.m_spend_public_key ||
```
Comment (vtnerd) [inline src/wallet/wallet2.cpp]: I do not understand the logic behind this check. `m_account` does not appear to be loaded from file, did I miss this? And if it was, is the goal to see if the same value differs in another location (i.e. the key is stored twice in the file)?

Comment (kenshi84) [inline src/wallet/wallet2.cpp]: I just copied the same check in the `THROW_WALLET_EXCEPTION_IF` below. This check makes sense to me, as `m_account` is filled in `load_keys(m_keys_file, password)` at line 2279. So it checks if the public wallet address recorded in the wallet cache is consistent with the keys.

**Comment by kenshi84** [comment]
@NanoAkron I posted [an issue there](https://github.com/USCiLab/cereal/issues/374) and @AzothAmmo replied that `boost::multi_index` may be supported in the future.

BTW, as long as we keep supporting the import of old boost serialized files, we'll still depend on boost.

**Comment by NanoAkron** [comment]
Great!

Well maybe we'll deprecate one day...would be nice to have everything truly cross-platform

**Comment by vtnerd** [comment]
Just to push some thoughts out here - I just checked and boost serialization has a portable archive in the example section which is missing float output. With some work it can added instead of Cereal. Its worth a consideration because the entire dependency on boost is unlikely to be removed anytime soon, so why bring an additional serialization library so similar ...

**Comment by vtnerd** [comment]
Also, the real portable test is going from a big endian to a little endian (or reverse) - anyone have access to such a box?

**Comment by kenshi84** [comment]
@vtnerd Could you post a link to such an example?

**Comment by kenshi84** [comment]
I created another PR #1462 inspired by @vtnerd's comment. Maybe better than cereal.

**Comment by fluffypony** [comment]
Closing in favour of #1462

**Pull request closed without merging**